flink-ml-lib/src/main/java/org/apache/flink/ml/stats/chisqtest/ChiSqTest.java

```diff
@@ -0,0 +1,492 @@
+                        .flatMap(new ExtractColAndFeatureAndLabel(inputCols, labelCol));
+
+        // compute the observed frequencies
+        DataStream<Tuple4<String, Object, Object, Long>> observedFreq =
+                DataStreamUtils.mapPartition(
+                        colAndFeatureAndLabel.keyBy(Tuple3::hashCode),
+                        new GenerateObservedFrequencies());
+
+        SingleOutputStreamOperator<Tuple4<String, Object, Object, Long>> filledObservedFreq =
+                observedFreq
+                        .transform(
+                                "filledObservedFreq",
+                                Types.TUPLE(
+                                        Types.STRING,
+                                        Types.GENERIC(Object.class),
+                                        Types.GENERIC(Object.class),
+                                        Types.LONG),
+                                new FillZeroFunc())
+                        .setParallelism(1);
+
+        // return a DataStream of the marginal sums of the factors
+        DataStream<Tuple3<String, Object, Long>> categoricalMargins =
+                DataStreamUtils.mapPartition(
+                        observedFreq.keyBy(tuple -> new Tuple2<>(tuple.f0, tuple.f1).hashCode()),
+                        new MapPartitionFunction<
+                                Tuple4<String, Object, Object, Long>,
+                                Tuple3<String, Object, Long>>() {
+                            @Override
+                            public void mapPartition(
+                                    Iterable<Tuple4<String, Object, Object, Long>> iterable,
+                                    Collector<Tuple3<String, Object, Long>> out) {
+                                HashMap<Tuple2<String, Object>, Long> map = new HashMap<>();
+
+                                for (Tuple4<String, Object, Object, Long> tuple : iterable) {
+                                    Long observedFreq = tuple.f3;
+                                    Tuple2<String, Object> key = new Tuple2<>(tuple.f0, tuple.f1);
+
+                                    if (map.containsKey(key)) {
+                                        Long count = map.get(key);
+                                        map.put(key, count + observedFreq);
+                                    } else {
+                                        map.put(key, observedFreq);
+                                    }
+                                }
+
+                                for (Tuple2<String, Object> key : map.keySet()) {
+                                    Long categoricalMargin = map.get(key);
+                                    out.collect(new Tuple3<>(key.f0, key.f1, categoricalMargin));
+                                }
+                            }
+                        });
+
+        // return a DataStream of the marginal sums of the labels
+        DataStream<Tuple3<String, Object, Long>> labelMargins =
```

Comment:
Same as above.

Reply:
As above, I used transform() and BoundedOneInput#endInput() to optimize the implementation.